src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -716,6 +860,18 @@ private void CreateFirstSegment(string segmentText)
                 throw ExceptionUtil.ResourceNotFoundError(ODataErrorStrings.RequestUriProcessor_CountOnRoot);
             }
 
+            if (this.IdentifierIs(UriQueryConstants.FilterSegment, identifier))
+            {
+                // $filter on root: throw
+                throw ExceptionUtil.ResourceNotFoundError(ODataErrorStrings.RequestUriProcessor_FilterOnRoot);
+            }
+
+            if (this.IdentifierIs(UriQueryConstants.EachSegment, identifier))
+            {
+                // $each on root: throw
+                throw ExceptionUtil.ResourceNotFoundError(ODataErrorStrings.RequestUriProcessor_EachOnRoot);
+            }
+

```

Comment:
should you also check this.IdentifierIs(UriQueryConstants.RefSegment, identifier) here?

Reply:
Yep, thanks for catching that.